lib/compress/zstd_lazy.c

```diff
@@ -531,8 +564,55 @@ size_t ZSTD_HcFindBestMatch_generic (
         matchIndex = NEXT_IN_CHAIN(matchIndex, chainMask);
     }
 
-    if (dictMode == ZSTD_dictMatchState) {
-        const ZSTD_matchState_t* const dms = ms->dictMatchState;
+    if (dictMode == ZSTD_dedicatedDictSearch) {
+        const U32 ddsChainSize         = (1 << dms->cParams.chainLog);
+        const U32 ddsChainMask         = ddsChainSize - 1;
+        const U32 ddsLowestIndex       = dms->window.dictLimit;
+        const BYTE* const ddsBase      = dms->window.base;
+        const BYTE* const ddsEnd       = dms->window.nextSrc;
+        const U32 ddsSize              = (U32)(ddsEnd - ddsBase);
+        const U32 ddsIndexDelta        = dictLimit - ddsSize;
+        const U32 ddsMinChain = ddsSize > ddsChainSize ? ddsSize - ddsChainSize : 0;
+        const U32 bucketSize           = (1 << ZSTD_LAZY_DDSS_BUCKET_LOG);
+        U32 attemptNb;
+
+        matchIndex = dms->hashTable[ddsIdx];
+
+        /* Empty chain */
+        if (!matchIndex)
+            return ml;
+
+        for (attemptNb = 0; attemptNb < bucketSize; attemptNb++) {
+            PREFETCH_L1(ddsBase + dms->hashTable[ddsIdx + attemptNb]);
+        }
+
+        for (attemptNb = 1; (matchIndex>ddsLowestIndex) & (nbAttempts>0) ; nbAttempts--, attemptNb++) {
+            size_t currentMl=0;
+            const BYTE* const match = ddsBase + matchIndex;
+            assert(match+4 <= ddsEnd);
+            if (MEM_read32(match) == MEM_read32(ip)) {
+                /* assumption : matchIndex <= dictLimit-4 (by table construction) */
+                currentMl = ZSTD_count_2segments(ip+4, match+4, iLimit, ddsEnd, prefixStart) + 4;
+            }
+
+            /* save best solution */
+            if (currentMl > ml) {
+                ml = currentMl;
+                *offsetPtr = current - (matchIndex + ddsIndexDelta) + ZSTD_REP_MOVE;
+                if (ip+currentMl == iLimit) break; /* best possible, avoids read overflow on next attempt */
+            }
+
+            if (matchIndex <= ddsMinChain) {
+                break;
+            }
+
+            if (attemptNb < bucketSize) {
```

Comment:
I wonder if it would be better to have two loops. One for the in-bucket attempts and one for the out of bucket attempts?

I guess it depends on how main chains fall out of the bucket.